## **PREDICTING THE 'GAS TURBINE DEGRADATION STATE COEFFICIENT'**

# **DATA DESCRIPTION**

The experiments have been carried out by means of a numerical simulator of a naval vessel (Frigate) characterized by a Gas Turbine (GT) propulsion plant. The different blocks forming the complete simulator (Propeller, Hull, GT, Gear Box and Controller) have been developed and fine tuned over the year on several similar real propulsion plants. In view of these observations the available data are in agreement with a possible real vessel.
In this release of the simulator it is also possible to take into account the performance decay over time of the GT components such as GT compressor and turbines.
The propulsion system behaviour has been described with this parameters:
- Ship speed (linear function of the lever position lp).
- Compressor degradation coefficient kMc.
- Turbine degradation coefficient kMt.


so that each possible degradation state can be described by a combination of this triple (lp,kMt,kMc).

# **DATA HANDLING**

In [ ]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt

In [ ]:
from google.colab import files
uploaded = files.upload()

Saving naval_vessel.csv to naval_vessel.csv


In [ ]:
 naval_vessel = pd.read_csv('naval_vessel.csv')
 naval_vessel

1.1380000  3.0000000  289.9640000  ...  0.0820000  0.9500000  0.9750000
0          2.088        6.0     6960.180  ...      0.287       0.95      0.975
1          3.144        9.0     8379.229  ...      0.259       0.95      0.975
2          4.161       12.0    14724.395  ...      0.358       0.95      0.975
3          5.140       15.0    21636.432  ...      0.522       0.95      0.975
4          6.175       18.0    29792.731  ...      0.708       0.95      0.975
...          ...        ...          ...  ...        ...        ...        ...
11928      5.140       15.0    21624.934  ...      0.471       1.00      1.000
11929      6.175       18.0    29763.213  ...      0.647       1.00      1.000
11930      7.148       21.0    39003.867  ...      0.834       1.00      1.000
11931      8.206       24.0    50992.579  ...      1.149       1.00      1.000
11932      9.300       27.0    72775.130  ...      1.704       1.00      1.000

[11933 rows x 18 columns]

In [ ]:
naval_vessel.columns = ['Lever position','Ship speed','Gas Turbine (GT) shaft torque','GT rate of revolutions','Gas Generator rate of revolutions','Starboard Propeller Torque','Port Propeller Torque','Hight Pressure (HP) Turbine exit temperature' ,'GT Compressor inlet air temperature','GT Compressor outlet air temperature','HP Turbine exit pressure' ,'GT Compressor inlet air pressure','GT Compressor outlet air pressure' ,'GT exhaust gas pressure','Turbine Injecton Control','Fuel flow','GT Compressor decay state coefficient','GT Turbine decay state coefficient']
naval_vessel.head(2)

Lever position  ...  GT Turbine decay state coefficient
0           2.088  ...                               0.975
1           3.144  ...                               0.975

[2 rows x 18 columns]

In [ ]:
naval_vessel.shape

(11933, 18)

In [ ]:
naval_vessel.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 11933 entries, 0 to 11932
Data columns (total 18 columns):
 #   Column                                        Non-Null Count  Dtype  
---  ------                                        --------------  -----  
 0   Lever position                                11933 non-null  float64
 1   Ship speed                                    11933 non-null  float64
 2   Gas Turbine (GT) shaft torque                 11933 non-null  float64
 3   GT rate of revolutions                        11933 non-null  float64
 4   Gas Generator rate of revolutions             11933 non-null  float64
 5   Starboard Propeller Torque                    11933 non-null  float64
 6   Port Propeller Torque                         11933 non-null  float64
 7   Hight Pressure (HP) Turbine exit temperature  11933 non-null  float64
 8   GT Compressor inlet air temperature           11933 non-null  float64
 9   GT Compressor outlet air temperature          11933 non-null 

In [ ]:
naval_vessel.dtypes

In [ ]:
naval_vessel.describe()

Lever position  ...  GT Turbine decay state coefficient
count    11933.000000  ...                        11933.000000
mean         5.167004  ...                            0.987501
std          2.626239  ...                            0.007500
min          1.138000  ...                            0.975000
25%          3.144000  ...                            0.981000
50%          5.140000  ...                            0.988000
75%          7.148000  ...                            0.994000
max          9.300000  ...                            1.000000

[8 rows x 16 columns]

In [ ]:
naval_vessel.value_counts

In [ ]:
naval_vessel.isnull().sum()

In [ ]:
#lambda function : For getting unique values for the columns we pass an anonymous function to apply().
print(naval_vessel.apply(lambda col: col.unique()))

We use lambda functions when we require a nameless function for a short period of time. In Python, we generally use it as an argument to a higher-order function (a function that takes in other functions as arguments). Lambda functions are used along with built-in functions like filter() , map() etc.

axis : {0 or 'index', 1 or 'columns'}, default 0
    Whether to drop labels from the index (0 or 'index') or
columns (1 or 'columns').

In [ ]:
##drop columns : GT compressor inlet air temperature , GT compressor inlet air pressure have only one unique value and doesn't contribute to our dataset.
naval_vessel = naval_vessel.drop('GT Compressor inlet air temperature', axis=1)
naval_vessel = naval_vessel.drop('GT Compressor inlet air pressure', axis =1)

In [ ]:
naval_vessel.head(3)

Lever position  ...  GT Turbine decay state coefficient
0           2.088  ...                               0.975
1           3.144  ...                               0.975
2           4.161  ...                               0.975

[3 rows x 16 columns]

# **EXPLORATORY DATA ANALYSIS (EDA)**

In [ ]:
import seaborn as sns

pairplot :This is a very simple command that will output a very handy graph showing the relationships between each of our numerical features by pairing each feature with all other features.

In [ ]:
sns.pairplot(naval_vessel ,size =15)

Output hidden; open in https://colab.research.google.com to view.

# **CORRELATION**

In [ ]:
correlation = naval_vessel.corr()

In [ ]:
print(correlation['GT Compressor decay state coefficient'])

Lever position                                 -2.183787e-04
Ship speed                                     -2.205444e-04
Gas Turbine (GT) shaft torque                   2.805532e-03
GT rate of revolutions                          1.225436e-03
Gas Generator rate of revolutions              -1.904056e-02
Starboard Propeller Torque                      5.975970e-04
Port Propeller Torque                           5.975970e-04
Hight Pressure (HP) Turbine exit temperature   -3.985651e-02
GT Compressor inlet air temperature                      NaN
GT Compressor outlet air temperature           -4.737307e-02
HP Turbine exit pressure                        8.004075e-03
GT Compressor inlet air pressure                3.085302e-12
GT Compressor outlet air pressure               8.160001e-03
GT exhaust gas pressure                         3.514751e-02
Turbine Injecton Control                       -3.218792e-02
Fuel flow                                      -1.383300e-02
GT Compressor decay stat

Heatmap : Correlation heatmap is graphical representation of correlation matrix representing correlation between different variables.

In [ ]:
correlation = np.round(naval_vessel.corr(),4)
plt.figure(figsize = (20,10))
sns.heatmap  (correlation, annot = True) 
#annot: writes the data value in each cell,an array-like with the same shape as data, we use this to annotate the heatmap instead of the data.
plt.show()

# **ALGORITHM**

In [ ]:
 X = naval_vessel[['Lever position','Ship speed','Gas Turbine (GT) shaft torque','GT rate of revolutions','Gas Generator rate of revolutions','Starboard Propeller Torque','Port Propeller Torque','Hight Pressure (HP) Turbine exit temperature','GT Compressor outlet air temperature','HP Turbine exit pressure','GT Compressor outlet air pressure' ,'GT exhaust gas pressure','Turbine Injecton Control','Fuel flow']]
 Y = naval_vessel['GT Compressor decay state coefficient']
 

In [ ]:
from sklearn.model_selection import train_test_split
X1_train,X1_test,Y1_train,Y1_test = train_test_split(X,Y,test_size = 0.20,train_size=0.80)
X2_train,X2_test,Y2_train,Y2_test = train_test_split(X,Y,test_size = 0.25 ,train_size=0.75)
X3_train,X3_test,Y3_train,Y3_test = train_test_split(X,Y,test_size = 0.30 ,train_size=0.70)
X4_train,X4_test,Y4_train,Y4_test = train_test_split(X,Y,test_size = 0.35 ,train_size=0.65)
X5_train,X5_test,Y5_train,Y5_test = train_test_split(X,Y,test_size = 0.40 ,train_size=0.60)

In [ ]:
from sklearn.linear_model import LinearRegression
model = LinearRegression(fit_intercept=True)
model.fit(X1_train, Y1_train)
model.fit(X2_train, Y2_train)
model.fit(X3_train, Y3_train)
model.fit(X4_train, Y4_train)
model.fit(X5_train, Y5_train)


LinearRegression()

In [ ]:
Y1_pred = model.predict(X1_test)
Y2_pred = model.predict(X2_test)
Y3_pred = model.predict(X3_test)
Y4_pred = model.predict(X4_test)
Y5_pred = model.predict(X5_test)

# **EVALUATION**

**Accuracy**

In [ ]:
accuracy_1 = print(model.score(X1_test, Y1_test))
accuracy_2 = print(model.score(X2_test, Y2_test))
accuracy_3 = print(model.score(X3_test, Y3_test))
accuracy_4 = print(model.score(X4_test, Y4_test))
accuracy_5 = print(model.score(X5_test, Y5_test))

0.8394426815907687
0.8440213472039819
0.8490387021552004
0.8469764368613395
0.838629083004502


**Mean Square Error**

In [ ]:
from sklearn.metrics import mean_squared_error
MSE1 = print(mean_squared_error(Y1_pred,Y1_test,squared = False))
MSE2 = print(mean_squared_error(Y2_pred,Y2_test,squared = False))
MSE3 = print(mean_squared_error(Y3_pred,Y3_test,squared = False))
MSE4 = print(mean_squared_error(Y4_pred,Y4_test,squared = False))
MSE5 = print(mean_squared_error(Y5_pred,Y5_test,squared = False))

0.005914053885278276
0.005820593320994999
0.005777413118797585
0.005779637503572315
0.005919818269196872


**R Square**

In [ ]:
from sklearn.metrics import r2_score
R_square1 = print(r2_score(Y1_pred,Y1_test))
R_square2 = print(r2_score(Y2_pred,Y2_test))
R_square3 = print(r2_score(Y3_pred,Y3_test))
R_square4 = print(r2_score(Y4_pred,Y4_test))
R_square5 = print(r2_score(Y5_pred,Y5_test))

0.8074477183889999
0.8143370293607296
0.8231320797902697
0.8207585132927426
0.8095554997841783


**Adjusted R2**

In [ ]:
Adj_R2_1 = print(1 - (1-r2_score(Y1_test, Y1_pred)) * (len(Y)-1)/(len(Y)-X.shape[1]-1))
Adj_R2_2 = print(1 - (1-r2_score(Y2_test, Y2_pred)) * (len(Y)-1)/(len(Y)-X.shape[1]-1))
Adj_R2_3 = print(1 - (1-r2_score(Y3_test, Y3_pred)) * (len(Y)-1)/(len(Y)-X.shape[1]-1))
Adj_R2_4 = print(1 - (1-r2_score(Y4_test, Y4_pred)) * (len(Y)-1)/(len(Y)-X.shape[1]-1))
Adj_R2_5 = print(1 - (1-r2_score(Y5_test, Y5_pred)) * (len(Y)-1)/(len(Y)-X.shape[1]-1))

0.8399919074810129
0.8438381200568814
0.8488613688635552
0.8467966810395622
0.8384395215983989


# **INSIGHTS**

 Linear Regression with 84% seems to be the best model for prediction.